app/src/main/kotlin/com/wire/android/ui/common/LinkifyText.kt

```diff
@@ -0,0 +1,245 @@
+//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//SOFTWARE.
+
+package com.wire.android.ui.common
+
+import android.os.Build
+import android.text.SpannableString
+import android.text.style.URLSpan
+import android.text.util.Linkify
+import androidx.compose.foundation.gestures.detectTapGestures
+import androidx.compose.material.LocalTextStyle
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.platform.LocalUriHandler
+import androidx.compose.ui.text.AnnotatedString
+import androidx.compose.ui.text.SpanStyle
+import androidx.compose.ui.text.TextLayoutResult
+import androidx.compose.ui.text.TextStyle
+import androidx.compose.ui.text.buildAnnotatedString
+import androidx.compose.ui.text.font.FontFamily
+import androidx.compose.ui.text.font.FontStyle
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.text.style.TextDecoration
+import androidx.compose.ui.text.style.TextOverflow
+import androidx.compose.ui.unit.TextUnit
+
+@Composable
+fun LinkifyText(
+    text: String,
+    mask: Int = Linkify.ALL,
+    modifier: Modifier = Modifier,
+    linkColor: Color = Color.Blue,
+    linkEntire: Boolean = false,
+    color: Color = Color.Unspecified,
+    fontSize: TextUnit = TextUnit.Unspecified,
+    fontStyle: FontStyle? = null,
+    fontWeight: FontWeight? = null,
+    fontFamily: FontFamily? = null,
+    letterSpacing: TextUnit = TextUnit.Unspecified,
+    textDecoration: TextDecoration? = null,
+    textAlign: TextAlign? = null,
+    lineHeight: TextUnit = TextUnit.Unspecified,
+    overflow: TextOverflow = TextOverflow.Clip,
+    softWrap: Boolean = true,
+    maxLines: Int = Int.MAX_VALUE,
+    onTextLayout: (TextLayoutResult) -> Unit = {},
+    style: TextStyle = LocalTextStyle.current,
+    clickable: Boolean = true,
+    onClickLink: ((linkText: String) -> Unit)? = null
+) {
+    val uriHandler = LocalUriHandler.current
+    val linkInfos = if (linkEntire) listOf(LinkInfo(text, 0, text.length)) else SpannableStr.getLinkInfos(text, mask)
+    val annotatedString = buildAnnotatedString {
+        append(text)
+        linkInfos.forEach {
+            addStyle(
+                style = SpanStyle(
+                    color = linkColor,
+                    textDecoration = TextDecoration.Underline
+                ),
+                start = it.start,
+                end = it.end
+            )
+            addStringAnnotation(
+                tag = "tag",
+                annotation = it.url,
+                start = it.start,
+                end = it.end
+            )
+        }
+    }
+    if (clickable) {
```

Comment:
I feel having 3 layers of nested objects with `LinkifyText`, `ClickableText` and `Text` is a bit too much no? To me `Linkify` and `ClickableText` is about the same thing. Maybe we could just get rid of one of them and just add the clickable overlay when needed (`clickable == true`). Otherwise it seems like a lot of extra boilerplate code we are adding here.

Reply:
well it's always needs

Reply:
What I meant is to just have the `LinkifyText` and a nested `Text` composable inside that checks the `clickable` flag to apply the appropriate `onTextLayout`. That way we would save quite a few loc. I tried the following and it still behaved correctly:

```@Composable
fun LinkifyText(
    text: String,
    mask: Int = Linkify.ALL,
    modifier: Modifier = Modifier,
    linkColor: Color = Color.Blue,
    linkEntire: Boolean = false,
    color: Color = Color.Unspecified,
    fontSize: TextUnit = TextUnit.Unspecified,
    fontStyle: FontStyle? = null,
    fontWeight: FontWeight? = null,
    fontFamily: FontFamily? = null,
    letterSpacing: TextUnit = TextUnit.Unspecified,
    textDecoration: TextDecoration? = null,
    textAlign: TextAlign? = null,
    lineHeight: TextUnit = TextUnit.Unspecified,
    overflow: TextOverflow = TextOverflow.Clip,
    softWrap: Boolean = true,
    maxLines: Int = Int.MAX_VALUE,
    onTextLayout: (TextLayoutResult) -> Unit = {},
    style: TextStyle = LocalTextStyle.current,
    clickable: Boolean = true,
    onClickLink: ((linkText: String) -> Unit)? = null
) {
    val uriHandler = LocalUriHandler.current
    val linkInfos = if (linkEntire) listOf(LinkInfo(text, 0, text.length)) else SpannableStr.getLinkInfos(text, mask)
    val annotatedString = buildAnnotatedString {
        append(text)
        linkInfos.forEach {
            addStyle(
                style = SpanStyle(
                    color = linkColor,
                    textDecoration = TextDecoration.Underline
                ),
                start = it.start,
                end = it.end
            )
            addStringAnnotation(
                tag = "tag",
                annotation = it.url,
                start = it.start,
                end = it.end
            )
        }
    }

    val onLinkClick = { offset: Int ->
        annotatedString.getStringAnnotations(
            start = offset,
            end = offset,
        ).firstOrNull()?.let { result ->
            if (linkEntire) {
                onClickLink?.invoke(annotatedString.substring(result.start, result.end))
            } else {
                uriHandler.openUri(result.item)
                onClickLink?.invoke(annotatedString.substring(result.start, result.end))
            }
        }
    }

    val layoutResult = remember { mutableStateOf<TextLayoutResult?>(null) }
    val pressIndicator = Modifier.pointerInput(onLinkClick) {
        detectTapGestures { pos ->
            layoutResult.value?.let { layoutResult ->
                onLinkClick(layoutResult.getOffsetForPosition(pos))
            }
        }
    }
    Text(
        text = annotatedString,
        modifier = modifier.then(pressIndicator),
        color = color,
        fontSize = fontSize,
        fontStyle = fontStyle,
        fontWeight = fontWeight,
        fontFamily = fontFamily,
        letterSpacing = letterSpacing,
        textDecoration = textDecoration,
        textAlign = textAlign,
        lineHeight = lineHeight,
        overflow = overflow,
        softWrap = softWrap,
        maxLines = maxLines,
        onTextLayout = {
            if (clickable) {
                layoutResult.value = it
                onTextLayout(it)
            }
        },
        style = style
    )
}